feat(terminal): add background command tracking and bug fixes for terminals

arafatkatze opened the pull request:
The entire list of bugs + changes covered in this PR is [mentioned here ](https://docs.google.com/document/d/1F1pQ-L1JljwJQ2JntTeXJH29Yr8tdNB930wk-ipwra4/edit?usp=sharing)

This PR enhances the terminal execution system with several improvements:

__1. Background Command Tracking__ When users click "Proceed While Running" on long-running commands (like `npm run dev`), the command is now tracked in the background with output logged to a temp file. This allows users to:

- Continue working while commands run
- Access full output via clickable log file links in the chat
- See a summary of running background commands in environment details

__2. Process Group Termination__ Fixed an issue where cancelling commands like `npm run dev` would only kill the parent shell, leaving child processes (webpack, node, etc.) running as orphans. Now uses:

- Unix: Process groups (`detached: true` + `process.kill(-pid)`) to kill entire process tree
- Windows: `taskkill /T /F` to kill process tree

__3. Memory Protection__ Added safeguards against memory exhaustion from commands with huge output:

- Caps `fullOutput` at 10MB, keeping the last 5MB when exceeded
- Truncates output to first/last 100 lines when exceeding 500 lines
- File-based logging for output exceeding 1000 lines or 512KB

__4. Centralized Constants__ Moved all terminal-related constants to `src/integrations/terminal/constants.ts` for better maintainability.

### Test Procedure

- Tested `npm run dev` cancellation - child processes now properly terminate
- Tested "Proceed While Running" - output logs to file, link is clickable
- Tested large output commands - memory stays bounded, output truncates gracefully
- JetBrains integration tests passing

### Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [x] ♻️ Refactor Changes

### Pre-flight Checklist

- [x] Changes are limited to a single feature, bugfix or chore
- [x] Tests are passing (`npm test`)
- [ ] I have created a changeset using `npm run changeset`
- [x] I have reviewed contributor guidelines

<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Enhances terminal command execution with background tracking, process management improvements, memory protection, and UI updates for better user interaction.
> 
>   - **Behavior**:
>     - Adds background command tracking in `CommandExecutor` and `StandaloneTerminalManager`, allowing commands to run in the background with output logged to temp files.
>     - Implements process group termination in `StandaloneTerminalProcess` to kill entire process trees on cancellation.
>     - Introduces memory protection in `VscodeTerminalProcess` and `StandaloneTerminalProcess` by capping `fullOutput` at 10MB and truncating output when exceeding 500 lines.
>     - Centralizes terminal-related constants in `constants.ts`.
>   - **UI**:
>     - Updates `ChatRow.tsx` to display clickable log file paths for background command outputs.
>     - Adds "Proceed While Running" button in `buttonConfig.ts` for command outputs.
>   - **Misc**:
>     - Fixes process termination logic using `terminateProcessTree` in `process-termination.ts`.
>     - Renames "Not Executed" to "Skipped" in `ChatRow.tsx`.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=cline%2Fcline&utm_source=github&utm_medium=referral)<sup> for 8bff366de5b47eca7e456ae5a4fd976e538c280d. You can [customize](https://app.ellipsis.dev/cline/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

abeatrix commented on src/integrations/terminal/standalone/StandaloneTerminalProcess.ts:
```diff
@@ -305,40 +319,56 @@ export class StandaloneTerminalProcess extends EventEmitter<TerminalProcessEvent
 		const pid = this.childProcess.pid
-		console.log(`[StandaloneTerminal] Terminating process ${pid} with SIGTERM`)
+		if (!pid) {
+			return
```
Should we try to kill it still?

```suggestion
			this.childProcess.kill("SIGTERM")
			return
```

abeatrix commented on src/integrations/terminal/CommandOrchestrator.ts:
```diff
@@ -162,31 +211,134 @@ export async function orchestrateCommandExecution(
+
+		// Write all existing lines to file
+		for (const line of outputLines) {
+			largeOutputLogStream.write(line + "\n")
```
while #1 is true, i don't think the #2 point is correct as we are calling "write()" on EVERY new line that comes in after the initial loop so even if it's handled by the internal buffering i think the overhead still exists ?

For #3,  I think joining 1000 lines into one string (which gabage collector will clean it up immediately) then make a write call once, is better than making 1000 write calls with stream buffer overhead. Plus at the 1000-line threshold (default MAX_LINES_BEFORE_FILE), we're talking ~100KB tops, so i don't think memory is an issue here.

arafatkatze commented on src/integrations/terminal/standalone/StandaloneTerminalProcess.ts:
```diff
@@ -176,37 +184,28 @@ export class StandaloneTerminalProcess extends EventEmitter<TerminalProcessEvent
 		const isCompiling =
-			compilingMarkers.some((marker) => data.toLowerCase().includes(marker.toLowerCase())) &&
-			!markerNullifiers.some((nullifier) => data.toLowerCase().includes(nullifier.toLowerCase()))
+			COMPILING_MARKERS.some((marker) => data.toLowerCase().includes(marker.toLowerCase())) &&
```
Added a [common function](https://github.com/cline/cline/blob/5c8e4190213ad557a63637790c85c09d1a4f52d7/src/integrations/terminal/constants.ts#L111) to check the compilation logic and made it such that we lower the false positives. Its now used both in Vscode and standalone terminal.

arafatkatze commented:
@saoudrizwan 
Thanks for the thorough review! Here are my responses to each point:

### 1. `isCompilingOutput()` Behavioral Change

**Resolved.** I've reverted this back to the original `includes()` behavior. 

The original intent was to reduce false positives (as Beatrix pointed out), but I agree we shouldn't cluster too many changes into one PR. I've kept the refactor (extracting it into a separate function with the marker arrays as constants) so it's easier to improve in a follow-up PR if needed.

```typescript
// Current (reverted to original behavior):
export function isCompilingOutput(data: string): boolean {
    const lowerData = data.toLowerCase()
    const hasMarker = COMPILING_MARKERS.some((marker) => lowerData.includes(marker.toLowerCase()))
    const hasNullifier = COMPILING_NULLIFIERS.some((nullifier) => lowerData.includes(nullifier.toLowerCase()))
    return hasMarker && !hasNullifier
}
```

### 2. Missing `lastRetrievedIndex` Update in `handleOutput`

**Intentional and correct.** The difference between VSCode and Standalone is due to different output retrieval patterns:

- **VSCode**: Uses polling-based retrieval. `getUnretrievedOutput()` is called periodically for busy terminals. Each poll should return only NEW output, so `lastRetrievedIndex` is updated as lines are emitted.

- **Standalone**: Uses event-based streaming. Output flows through events to CommandOrchestrator, which collects it. `getUnretrievedOutput()` is only called AFTER `emitRemainingBuffer()` has already run (on completion or "Proceed While Running"), which updates `lastRetrievedIndex`.

The key is that `emitRemainingBuffer()` (called from `continue()` and process close) updates `lastRetrievedIndex = fullOutput.length`, marking the boundary before any new output arrives. This ensures no duplicates.

### 3. No Fallback When `treeKill` Fails

**Resolved.** I've added a proper SIGKILL fallback with a 2-second timeout and extracted it into a reusable utility function:

```typescript
// src/utils/process-termination.ts
export async function terminateProcessTree(options: TerminateProcessTreeOptions): Promise<void> {
    const { pid, childProcess, isCompleted, gracefulTimeoutMs = 2000 } = options

    // Send SIGTERM for graceful shutdown
    treeKill(pid, "SIGTERM")

    // Wait for graceful shutdown or timeout
    const gracefulTimeout = new Promise<void>((resolve) => setTimeout(resolve, gracefulTimeoutMs))
    const processExit = new Promise<void>((resolve) => {
        if (childProcess) {
            childProcess.once("exit", () => resolve())
        } else {
            resolve()
        }
    })

    await Promise.race([processExit, gracefulTimeout])

    // Force kill if still running
    if (!isCompleted()) {
        treeKill(pid, "SIGKILL")
    }
}
```
